capa/ghidra/capa_explorer.py

```diff
@@ -0,0 +1,242 @@
+# Integrate capa results with Ghidra UI
+# @author Colton Gabertan (gabertan.colton@gmail.com)
+# @category Python 3.capa
+
+# Copyright (C) 2023 Mandiant, Inc. All Rights Reserved.
+# Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at: [package root]/LICENSE.txt
+# Unless required by applicable law or agreed to in writing, software distributed under the License
+#  is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and limitations under the License.
+import sys
+import json
+import logging
+import pathlib
+from typing import Dict, List
+from contextlib import suppress
+
+from ghidra.app.cmd.label import CreateNamespacesCmd
+from ghidra.program.model.symbol import Namespace, SourceType, SymbolType
+
+import capa
+import capa.main
+import capa.rules
+import capa.render.json
+import capa.ghidra.helpers
+import capa.features.extractors.ghidra.extractor
+
+logger = logging.getLogger("capa_explorer")
+
+
+class CAPADATA:
+    def __init__(self, namespace, scope, capability, locations, node, attack=None):
+        self.namespace = namespace
+        self.scope = scope
+        self.capability = capability
+        self.locations = locations
+        self.node = node
+        self.attack = attack
+
+    def recurse_node(self, node_dict):
+        """pull string data by recursing node dicts"""
+
+        if not node_dict:
```

Comment:
Do we expect `node_dict` to be `None` or an empty Python Dictionary?

Reply:
So there's a few cases in which a node doesn't necessarily have a description that we should write to the UI.

For example, a previous call may recurse on a `node_dict` like `{'type': 'subscope', 'scope':'basicblock'}. The recursive search relies on the `'type'` key's value to be the key for the description we want to print. An example of what the function expects is something like `node_dict = {'type': 'api', 'api': 'RegisterServiceCtrlHandler'}`.

So in essence, this check just handles key mismatches.

Reply:
Thanks for the catch. To be more explicit, I've changed to `if node_data is None`, since a key mismatch would return NoneTypes. All final data from the recursion should be either a `str` or an `int` that was casted to a `str` via the `hex()` call.